text/0000-deprecation-warning-handlers.md

```diff
@@ -0,0 +1,172 @@
+handler would look like:
+
+```js
+import { registerHandler } from "ember-debug/deprecations";
+
+registerHandler(function deprecationHandler(message, options) {
+  // * message is the deprecation message
+  // * options.until is the version this deprecation will be removed at
+  // * options.id is the canonical id for this deprecation
+  if (options.until === "2.4.0") {
+    throw new Error(message);
+  } else {
+    console.log(message);
+  }
+});
+```
+
+Warnings are similar, but will not recieve an `until` value:
+
+```js
+import { registerHandler } from "ember-debug/warnings";
+
+registerHandler(function warningHandler(message, options) {
+  // * message is the warning message
+  // * options.id is the canonical id for this warning
+  if (options.id !== 'view.rerender-on-set') {
+    console.log(message);
+  }
+});
+```
+
+##### chained handlers
+
+Since several handlers may be registered, a method of deferring to a previously
+registered handler must be allowed. A third option is passed to handlers, the
+function `next` which represents the previously registered handler.
+
+For example:
+
+```js
+import { registerHandler } from "ember-debug/deprecations";
+
+registerHandler(function firstDeprecationHandler(message, options, next) {
+  console.warn(message);
+});
+
+registerHandler(function secondDeprecationHandler(message, options, next) {
+  if (options.until === "2.4.0") {
+    throw new Error(message);
+  }
+  next(...arguments);
+});
+```
+
+The first registed handler will receive Ember's default behavior as `next`.
+
+##### new assertions for deprecate and warn
+
+Ember's APIs for deprecation and warning do not currently require any information
+beyond a message. It is proposed that deprecations be **required** to pass
+the following information:
+
+* Message
+* Test
+* Canonical id (with a format of `package-name.some-id`)
+* Release when this deprecation will be stripped
+
+For example:
+
+```
+import Ember from "ember";
+
+Ember.deprecate("Some message", false, {
+  id: 'ember-routing.query-params',
+  until: '3.0.0'
+});
+```
+
+If this information is not present and assertion will be made.
+
+Warnings likewise will be required to pass a canonical id:
+
+```
+import Ember from "ember";
+
+Ember.warn("Some warning", {id: 'ember-debug.something'});
+```
+
+##### default handlers
+
+The default handler for deprecation should be quite simple, and mirrors current
+behavior:
+
+```js
+function defaultDeprecationHandler(message, options) {
+  if (Ember.ENV.RAISE_ON_DEPRECATION) {
+   throw new Error(format(message, options));
+  } else {
+   console.log(format(message, options));
+  }
+}
+```
+
+The default handler for warnings would be simple `console.log`.
+
+# Drawbacks
+
+By not providing a robust log-level API, we are punting complexity to the
+consumer of this API. For a low-level tooling API such as this one, it seems
+and appropriate tradeoff.
+
+# Alternatives
+
+Each app can stub out `deprecate` and `warn`.
```

Comment:
Can the template complier also have configuration to stub `deprecate` and `warn` too? seems like it would be a big advantage for example that would be ideal compared with debugging the asset pipeline.